Use goto_functiont::parameter_identifiers and remove goto_functiont::type

tautschnig opened the pull request:
Having two function type definitions (one in the symbol table and one in
goto_functiont) is just prone to become inconsistent. Instead, always look at
the symbol table when we do need type information. If all we care about are the
identifiers we can still use goto_functiont::parameter_identifiers.

~I'm also marking this RFC, @kroening in particular: it's quite painful to maintain consistency between `parameter_identifiers` and `code_typet::parameters()` - do we really need `parameter_identifiers`, or shouldn't we instead always use the information stored in the type?~

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
Also marking this do-not-review: I should try to break this up into a series of smaller commits that prepare the final change.

kroening commented:
@tautschnig For a long list of reasons, it's best to get rid of the identifiers in `code_typet::parameters()` and only use the ones in `goto_functiont`.
I am happy to get rid of the signature in `goto_functiont`, for the reason that you are giving.

peterschrammel commented:
On a side note: I've seen an issue with function pointer types recently: i.e. a function pointer variable has the original return type whereas the symbol table has return type void for the pointed function after the remove_return pass. Type consistency validations fail due to that.

kroening commented:
On the side note: One cleanup item on my list is to change the way remove_returns works: the idea is to keep the return type as is. This removes the need to keep a copy of the original return type.

tautschnig committed:
Remove goto_functiont::type

Having two function type definitions (one in the symbol table and one in
goto_functiont) is just prone to become inconsistent. Instead, always look at
the symbol table when we do need type information. If all we care about are the
identifiers we can still use goto_functiont::parameter_identifiers.

tautschnig committed:
Do not read from goto_functiont::type

Future changes will remove this member as it is redundant with the information
stored in the symbol table. We keep writing to it so as not to break any users,
but no longer read it.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e8d7597a88a06cd5dc84ebb95a89c9122a5733d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102934571
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

kroening commented:
I would leave the note about `parameter_identifiers` around for a while, say until `codet::parametert::get_identifier()` is gone.

tautschnig commented:
> I would leave the note about `parameter_identifiers` around for a while, say until `codet::parametert::get_identifier()` is gone.

Done, and updated - assuming that was the comment you were after.

kroening commented:
Apologies for iterating on this PR.

The PR goes beyond the parameter identifiers, and also touches the "is hidden" feature. The question of whether or not the execution of a functions should be hidden is a property of the body (it's triggered via a label in the body), and not a property of the function signature.

It should therefore be in goto_functiont, and not the type.

martin-cs commented:
@tautschnig Do you want this reviewed as is?

tautschnig commented:
> Apologies for iterating on this PR.
> 
> The PR goes beyond the parameter identifiers, and also touches the "is hidden" feature. The question of whether or not the execution of a functions should be hidden is a property of the body (it's triggered via a label in the body), and not a property of the function signature.
> 
> It should therefore be in goto_functiont, and not the type.

That all makes sense, but really means I should first revert #4312. I'll create a PR to that effect.

tautschnig commented:
The "is hidden" PR is now posted as #5575.

tautschnig commented:
@kroening @martin-cs This is now rebased on top of the changes implementing "is hidden" as requested.

martin-cs commented on src/goto-programs/goto_function.h:
```diff
@@ -16,29 +16,23 @@ Date: May 2018
   ///
   /// Note: This is now the preferred way of getting the identifiers of the
-  /// parameters. The identifiers in the type will go away.
+  /// parameters. The identifiers in the type are now gone.
```
:smile: nice!

kroening replied:
I am afraid they are still there. "The type" here refers to the type in the symbol table.

tautschnig replied:
Err, true. Change to the comment reverted/dropped.

martin-cs reviewed (approved):
Seems good.
